include/srf/node/sink_channel_base.hpp

```diff
@@ -0,0 +1,237 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <srf/channel/buffered_channel.hpp>
+#include <srf/channel/ingress.hpp>
+#include <srf/constants.hpp>
+#include <srf/exceptions/runtime_error.hpp>
+#include <srf/node/edge.hpp>
+#include <srf/node/edge_properties.hpp>
+#include <srf/node/forward.hpp>
+#include <srf/node/sink_properties.hpp>
+#include <srf/utils/type_utils.hpp>
+#include "srf/channel/egress.hpp"
+
+#include <mutex>
+
+namespace srf::node {
+
+/**
+ * @brief Extends SinkProperties to hold a Channel and provide SinkProperties access the the channel ingress
+ *
+ * @tparam T
+ */
+template <typename T>
+class SinkChannelBase
+{
+  protected:
+    SinkChannelBase();
+
+  public:
+    /**
+     * @brief Enables persistence of the Channel.
+     *
+     * The default behavior is to close the owned Channel after all Ingress objects created from channel_ingress have
+     * been destroyed. The persistent option keeps an internal Ingress object live preventing channel closure.
+     *
+     * Invoking stop on the Runner which owns the Runnable explicitly disables persistence.
+     */
+    void enable_persistence();
+
+    /**
+     * @brief Disables persistence of the Channel
+     */
+    void disable_persistence();
+
+    /**
+     * @brief Determine if the Channel is persistent or not.
+     *
+     * @return true
+     * @return false
+     */
+    bool is_persistent() const;
+
+    /**
+     * @brief Replace the current Channel.
+     *
+     * An update can only occur if persistence has not be been requested and no external entities have acquired an
+     * Ingress.
+     *
+     * @param channel
+     */
+    void update_channel(std::unique_ptr<Channel<T>> channel);
+
+    /**
+     * @brief The number of outstanding edge connections from this SinkChannel to SourceChannels
+     *
+     * @return std::size_t
+     */
+    std::size_t use_count()
+    {
+        return m_ingress.use_count();
+    }
+
+  protected:
+    inline channel::Egress<T>& egress()
+    {
+        DCHECK(m_channel);
+        return *m_channel;
+    }
+
+    // access the full channel
+    std::shared_ptr<channel::Channel<T>> channel();
+
+    // alternative to update_channel
+    // disables ingress_channel
+    void set_shared_channel(std::shared_ptr<channel::Channel<T>> channel);
+
+    // thread safe access to
+    [[nodiscard]] std::shared_ptr<channel::Ingress<T>> ingress_channel();
+
+  private:
+    [[nodiscard]] std::shared_ptr<channel::Ingress<T>> ingress_channel_unsafe();
```

Comment:
What is the purpose of `ingress_channel` vs `ingress_channel_unsafe`? Is it just to avoid the deadlock in `enable_persistence`? If so, I feel like a `recursive_mutex` would be a better implementation.

Reply:
Specifically to avoid the recursive mutex.

Reply:
Whats wrong with a recursive mutex? Its not clear at all when you should call `ingress_channel_unsafe` vs `ingress_channel`